Hidden Markov models

rpauszek opened the pull request:
**Why this PR?**
Adds models for analyzing time series data exhibiting transitions between discrete states. This is useful for constant-extension experiments or photon count time traces (especially FRET)

This PR lays down all of the internal algorithms for model training and state labeling. A second PR will focus on building up the rest of the public API with convenience functions for plots, histograms, etc...

Notes:
- Similar to `lk.GaussianMixtureModel` that wraps `sklearn`'s GMM, I defined an internal model class to just handle the parameters and initialization/update; we can then expose what we want from the model through the public API that wraps this
- There is a lot of internal code that uses notation seemingly from nowhere, however it is pretty standard in the HMM field and derives largely from Rabiner's seminal "A tutorial on HMMs and selected applications in speech recognition". Another very useful resource is the [HMM project](https://web.ece.ucsb.edu/Faculty/Rabiner/ece259/speech%20recognition%20course/term%20projects/HMM%20Project.pdf) from his speech recognition course, although not exactly sure if we should or how we should cite this
- Mostly I did this to avoid using things like `state_transition_matrix` sprinkled in throughout the algorithm, instead using the standard `A`. Open for discussion if there are suggestions for finding a middle ground here.
- For the public properties I use more descriptive attribute names (`means` instead of `mu`)
- Much of the code is pulled from my `smtirf` repo, written during my postdoc and refactored to fit `pylake`'s API style and vectorized where possible (since we're not yet using `numba`)
- The only intended public API is the `HiddenMarkovModel` class, although I haven't made it public yet. My idea is to match the interface with that of `lk.GaussianMixtureModel`. 

A typical workflow (after completing the public API) would be something like:
```
model = lk.HiddenMarkovModel.from_channel(force1x, n_states=3)
model.plot(force1x)
```

<img width="1488" alt="image" src="https://github.com/lumicks/pylake/assets/61475504/4d84bd2f-368a-4509-92f5-f733023ede6c">

JoepVanlier commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,361 @@
+        gamma : np.ndarray
+            probability of being in state i at time t, shape [T, K]
+        xi : np.ndarray
+            probaility of being in state i at time t and state j at time t + 1, shape [T, K, K]
```
```suggestion
            probability of being in state i at time t and state j at time t + 1, shape [T, K, K]
```

JoepVanlier commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,361 @@
+
+    def __init__(self, data, n_states, tol, max_iter):
+        self.n_states = n_states
+        self._model, self._exit_flag = baum_welch(
```
I like the approach of separating the model that uses all the shorthand notation from the user API. I would probably move that model to the `detail` folder (similarly to how we've done for the FEC models) if it's not meant to be used by end users.

JoepVanlier commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,361 @@
+class ClassicHmm:
+    """Model parameters for classic Hidden Markov Model.
+
+    Parameters for this internal class conform to notation in Rabiner [ref]
```
Add ref.

JoepVanlier commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,361 @@
+        )
+
+    @property
+    def exit_flag(self):
+        return self._exit_flag
```
Personally, I find the use of `exit_flags` to indicate convergence failure fairly unpythonic.
I find this awkward with the GMM implementation we have as well.

Exit flags like this easily go unnoticed and long term, it would be nice to actually go to emit warnings for these.

Also, why not make these attributes full properties rather than sticking them in a dictionary?

JoepVanlier reviewed (changes requested):
1 of N:
Code-wise it looks nice. I like the separation of the internal model from the API. I also like that the model essentially is frozen, and how you set up the guess function to return one with an initial guess. The design I like.

What I'm really missing is tests though and this makes it a lot harder to review. Tests are usually quite helpful to see how you typically set up your models and simulations.

I've started reading the Rabiner paper, but I haven't had time yet to delve into the details. I have so far only commented a few concerns that popped out after an initial look over the code.

rpauszek commented:
Update Dec 8 2023:
@JoepVanlier  I've implemented your requested changes and added further convenience API to match that of our GMM (plotting, histograms, etc). There are a lot of small commits but I thought that might make it easier to review. If not I can squash down (I think this whole thing will end up as 1 commit in the end)

A few notes
- there's the fixup with the issue we ran into with the monkeypatching. I left that as a separate commit right at the beginning
- I forgot to move stuff to detail before making changes .... so I just did that in one go, literal copy/paste from one commit to the next
- rather than adding a `from_channel` class method like for the GMM I found it nicer to just deal with it in the constructor. I'm actually re-thinking if we even want to allow using these models with numpy arrays because a lot of the downstream stuff relies on time information. so to make it actually useful we'd need to either
   - allow users to optionally supply  the time step if they're using regular arrays
   - disallow this altogether
- I think I want to get rid of the `label` method for this and GMM and replace it either with `.path()` like I have here. Or might be more useful to have two separate methods `.state_path()` and `.emission_path()` But I'll do this in the next PR so this one doesn't go too crazy with adding functionality and changing existing API
- docs seem a bit incomplete, but I want to do another PR to make changes to the GMM API that will I think be nicer user experience in the end and avoids writing long/complex example code for certain visualizations I want to show for comparing between GMM and HMM

JoepVanlier commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,210 @@
+        The tolerance for training convergence.
+    max_iter : int
+        The maximum number of iterations to perform.
+    gmm_guess : GaussianMixtureModel
+        Initial guess for the observation model parameters.
```
Maybe not for this PR (it's big enough 😅) but it would be nice if we could pass in an hmm as well. If a previous fit didn't converge and it took a while, we could continue the run?

JoepVanlier reviewed (changes requested):
Excellent work. A few more comments. I'm not fully done with reviewing it yet, but it's getting there. 👍 

Personally, rather than wrap changes to the HMM on the next PR, I'd prefer to just have the changes that you would do to the HMM in this one already so that I can see what the API will look like.

JoepVanlier commented on docs/refs.bib:
```diff
@@ -445,3 +445,14 @@ @article{bullerjahn2020optimal
+  pages={257-289},
+  year={1989},
+  publisher={IEEE}
+}
```
eolrip :)

JoepVanlier commented on docs/tutorial/population_dynamics.rst:
```diff
@@ -133,6 +126,50 @@ two different states and do not arise from any (bio)physically relevant mechanis
+  we must infer the conformational state from some observed signal; ie., the force measurements.
+  The observed signal is described by a GMM.
+
+This description is the most basic implementation of an HMM. For detailed theoretical and practical
+explanations on the use and training of HMMs see the seminal work by Rabiner :cite:`Rabiner1989HMM`.
```
This is a nice way of dealing with not having a theory section on this yet.

JoepVanlier commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,210 @@
+        )
+
+    @property
+    def exit_flag(self):
```
Same as above. You put a lot more info in here than just a flag.

rpauszek commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,210 @@
+        """
+        return viterbi(trace.data, self._model)
+
+    def path(self, trace):
```
yeah, I kept going back and forth on the idea which is how I ended up here. but in the end I think it's best to have two separate methods

rpauszek commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,210 @@
+            Plotting keyword arguments passed to the data line plot.
+        path_kwargs : Optional[dict]
+            Plotting keyword arguments passed to the state path line plot.
+        show_path : bool
+            Whether or not to show state path.
```
at one point it did 😯

rpauszek commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,210 @@
+            PDF array split into components for each state with shape (n_states, x.size).
+            The full normalized PDF can be calculated by summing across rows.
+        """
+        return np.exp(self._model.state_log_likelihood(x))
```
yeah, I wasn't quite happy with this because technically the HMM doesn't have a PDF in the traditional sense I guess. Considering right now we only use it for the histogram and it's a single line of code, I'm gonna get rid of the method altogether. We can always add it back (private or otherwise) if we find we need it in other places in the future

rpauszek commented on lumicks/pylake/population/detail/hmm.py:
```diff
@@ -0,0 +1,296 @@
+
+            # check for convergence
+            if (delta := current_log_likelihood - previous_log_likelihood) < 0:
+                warnings.warn(f"log likelihood decreasing by {delta:0.4e}")
```
Yeah I'm not sure. I've seen this in other EM implementations but I think it's just a numerical wiggle as the EM algorithm guarantees that the likelihood increases with each iteration. No idea where this comes from, but in all the data I've tried it's always small, on the order of 1e-14 or so. Maybe best to just get rid of the warning altogether?

rpauszek commented on lumicks/pylake/population/detail/hmm.py:
```diff
@@ -0,0 +1,296 @@
+    statepath = np.zeros(T, dtype=int)
+
+    # initialization
+    delta = col(np.log(model.pi) + B[0])
```
yes, for very well separated states the initial state probability collapses to something like `[1, 0, 0]`. This is a peculiarity of training on a single trace since you only have one initial observation so you can't calculate a proper distribution. If you train on an ensemble of traces you would get an actual distribution. 

If you use data where the states are closer (or observations are noiser) you get something along the lines of `[1.28634832e-201 1.99176214e-049 1.00000000e+000]` which doesn't throw this warning. So I suggest to just ignore this warning using `np.errstate`

JoepVanlier replied:
Yeah, I think that is appropriate, since `exp(-huge)` will just result in zero.

rpauszek commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,210 @@
+
+        return state_path, emission_path
+
+    def extract_dwell_times(self, trace, *, exclude_ambiguous_dwells=True):
```
yes, but now getting rid of `label()` I use `state_path()` here but `label()` still in the GMM. I started moving stuff to a mixin class, but that'll now involve changes to the GMM and it got messy quickly. So I suggest leaving this (and `hist`) as it is right now and addressing that in the 2nd PR? that ok?

JoepVanlier commented on lumicks/pylake/population/detail/hmm.py:
```diff
@@ -0,0 +1,298 @@
+        delta = col(np.log(model.pi) + B[0])
+    # recursion
+    for t in range(1, T):
+        R = delta + np.log(model.A)
```
Not so likely for real data, but if a transition is never observed (and this element becomes zero), you could get a runtime warning taking the log here as well.

rpauszek replied:
actually it's a real case when two states aren't connected to each other, you get a `0` for that element. added handling for this and a test

JoepVanlier replied:
Excellent

JoepVanlier reviewed (changes requested):
Looking great! Aside from validating that the HMM or GMM initial guess is passed actually contains the correct number of states, it's mostly minor things.

rpauszek commented on lumicks/pylake/population/hmm.py:
```diff
@@ -0,0 +1,210 @@
+        return 1 / self._model.tau
+
+    @property
+    def std(self) -> np.ndarray:
```
copy/paste from GMM. I think Tobias added it there but nothing else has annotations. I'll leave it out for now, maybe worthwhile at some point doing a specific look through for everything to annotate where appropriate. Still not quite sure how I feel about them as I think it makes the code much less readable

JoepVanlier reviewed (approved):
:shipit:
